### PR TITLE
listener: implement disabled predicates

### LINKED ISSUE #9032: Add a port white list in sniffer-ish listener filter
*Description*:
A per listener time out doesn't work well in iptables-redirect-to listener.
For example:
For server-first -packet protocol filter chain: a immediate timeout is preferred
For TLS expected filter chain: it is expected to be infinity timeout

I propose adding a L4 address based timeout in sniffer listener filter (tls_inspector)
```
Timeouts: 
- matcher:
   port: 3306 #mysql server first protocol 
  timeout: 0.0000001s
- matcher: {} 
  timeout: 100000000s
```
Alternative solution:
  list of port for infinity timeout, and a list of port for immediate timeout

### PR DESCRIPTION
**Pull request by lambdai**
Description: 
Implement filer_disabled which is introduced in #9393
`filter_disabled` is the only common field in listener filter so there are some signature update on the filter factory.

Risk Level: LOW 
Testing: Integration test and unit tests
Docs Changes
Release Notes:
Fixes #9032
[Optional Deprecated:]

### CODE REVIEW

File: api/envoy/api/v2/listener/listener_components.proto [hunk 1]
```diff
@@ -255,17 +254,14 @@ message ListenerFilter {
...
-  //       rule.destination_port_range = Int32Range {start = 3306, end = 3307},
-  //       rule.destination_port_range = Int32Range {start = 15000, end = 15001},
-  //    ]
+  //
+  // Examples:
+  //
+  // * clear the field if filter is always enabled
+  // * disable when the destination port is 3306
+  //   ``rule.destination_port_range = Int32Range {start = 3306, end = 3307}``
+  // * disable when the destination port is 3306 or 15000
```
Comment (htuch) [inline api/envoy/api/v2/listener/listener_components.proto]: Random drive-by, looks like this is a bit of a regression in API docs, they've lost some examples.

Comment (lambdai) [reply api/envoy/api/v2/listener/listener_components.proto]: Thanks for the comment! I made this change because the # of lines of the comments is large IMHO. 
I feel the doc should describe the default behavior of the field and leave the examples to the message type doc.

I can revert if you prefer the verbose version before this PR

Comment (htuch) [reply api/envoy/api/v2/listener/listener_components.proto]: I think either get rid of the example and point to a full one in the docs or just provide some self-contained examples here. Having half formed examples isn't that great.

Comment (mattklein123) [reply api/envoy/api/v2/listener/listener_components.proto]: +1 to doing the full example in one place or the other. Please also add a release note in addition to any arch level docs.

File: include/envoy/network/filter.h [hunk 2]
```diff
@@ -274,6 +274,27 @@ class ListenerFilterCallbacks {
 /**
  * Listener Filter
  */
+
+//
```
Comment (mattklein123) [inline include/envoy/network/filter.h]: ?

File: include/envoy/network/filter.h [hunk 3]
```diff
@@ -301,7 +322,8 @@ class ListenerFilterManager {
    * first is called first).
    * @param filter supplies the filter being added.
    */
-  virtual void addAcceptFilter(ListenerFilterPtr&& filter) PURE;
+  virtual void addAcceptFilter(Network::ListenerFilterConfigSharedPtr listener_filter_config,
```
Comment (mattklein123) [inline include/envoy/network/filter.h]: Update docs

File: include/envoy/server/filter_config.h [hunk 4]
```diff
@@ -254,12 +254,14 @@ class NamedListenerFilterConfigFactory : public ListenerFilterConfigFactoryBase
    * of general error or a Json::Exception if the json configuration is erroneous. The returned
    * callback should always be initialized.
    * @param config supplies the general protobuf configuration for the filter
+   * @param
```
Comment (mattklein123) [inline include/envoy/server/filter_config.h]: ?

**Comment by lambdai** [comment]
Fixing...

File: api/envoy/api/v2/listener/listener_components.proto [hunk 5]
```diff
@@ -255,17 +278,8 @@ message ListenerFilter {
...
-  // 0. always enable filter
-  //    don't set `filter_disabled`
-  // 1. disable when the destination port is 3306
-  //    rule.destination_port_range = Int32Range {start = 3306, end = 3307}
-  // 2. disable when the destination port is 3306 or 15000
-  //    rule.or_match = MatchSet.rules [
-  //       rule.destination_port_range = Int32Range {start = 3306, end = 3307},
-  //       rule.destination_port_range = Int32Range {start = 15000, end = 15001},
-  //    ]
+  // Clear the field if filter is always enabled.
```
Comment (mattklein123) [inline api/envoy/api/v2/listener/listener_components.proto]: What do you mean by this? Can you rephrase to what this field does and what the default behavior is if it's empty?

File: source/common/network/filter_matcher.cc [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+
+#include "absl/strings/str_format.h"
+
+namespace Envoy {
+namespace Network {
+
+// Introduced to save number of memory allocation at build phase memory usage match phase.
+// Work with OwnedListenerFilterMatcher.
+// The first phase is constructor, requiring only 1 slots.
+// The second phase is complete() which reserves contiguously slots.
```
Comment (mattklein123) [inline source/common/network/filter_matcher.cc]: Sorry can't parse this. Can you expand/clarify? Also clarify below what complete() does. It will help me review.

File: source/common/network/filter_matcher.h [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+#include "envoy/network/filter.h"
+#include "envoy/network/listen_socket.h"
+
+namespace Envoy {
+namespace Network {
+
+/**
+ * A non-empty listener filter matcher which aggregates multiple sub matchers.
+ */
+class OwnedListenerFilterMatcher : public ListenerFilterMatcher {
```
Comment (mattklein123) [inline source/common/network/filter_matcher.h]: s/OwnedListenerFilterMatcher/SetListenerFilterMatcher (or similar)

File: source/server/connection_handler_impl.h [hunk 8]
```diff
@@ -229,9 +229,40 @@ class ConnectionHandlerImpl : public Network::ConnectionHandler,
...
+        if (matcher_ == nullptr) {
+          return false;
+        } else {
+          return matcher_->matches(cb);
+        }
+      }
+
+    private:
+      Network::ListenerFilterPtr listener_filter_;
+      Network::ListenerFilterMatcherSharedPtr matcher_;
```
Comment (mattklein123) [inline source/server/connection_handler_impl.h]: nit: const

File: include/envoy/network/filter.h [hunk 9]
```diff
@@ -271,6 +271,28 @@ class ListenerFilterCallbacks {
...
+  virtual ~ListenerFilterMatcher() = default;
+  virtual bool matches(Network::ListenerFilterCallbacks& cb) const PURE;
+};
+using ListenerFilterMatcherPtr = std::unique_ptr<ListenerFilterMatcher>;
+using ListenerFilterMatcherSharedPtr = std::shared_ptr<ListenerFilterMatcher>;
+
+/**
+ * A mapping from a general listener filter message without `typed_config`.
+ */
+class ListenerFilterConfig {
```
Comment (mattklein123) [inline include/envoy/network/filter.h]: Why is this interface needed? Can't the matcher itself just be passed to where it is needed (or null)?

File: source/common/network/filter_matcher.cc [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+  }
+  case envoy::config::listener::v3::ListenerFilterChainMatchPredicate::RuleCase::
+      kDestinationPortRange: {
+    return std::make_unique<DstPortMatcher>();
+  }
+  // Invalid message.
+  case envoy::config::listener::v3::ListenerFilterChainMatchPredicate::RuleCase::RULE_NOT_SET: {
+    throw EnvoyException(
+        absl::StrFormat("invalid listener filter chain matcher: %s", match_config.DebugString()));
+  }
```
Comment (mattklein123) [inline source/common/network/filter_matcher.cc]: This should be impossible via schema (make sure there is a validator). Same below. Use NOT_REACHED instead

File: include/envoy/network/filter.h (hunk 9 above)
Comment (lambdai) [reply include/envoy/network/filter.h]: For this PR, yes we can. 
However, I feel it's helpful to have common listener filter config besides a config for a particular config.
A reasonable example is that we planned to set timeout for individual listener filter(and avoid adding timeout() for each sniffer listener filter), instead of a listener time out for the entire listener.
WDYT? I can revert to pure matcher if you like. But I do think this ListenerFilterConfig is nice to have. Perhaps renaming to CommonListenerFilterConfig?

File: source/common/network/filter_matcher.cc (hunk 10 above)
Comment (lambdai) [reply source/common/network/filter_matcher.cc]: Sure. Will add validator if it doesn't exist yet.

File: source/server/connection_handler_impl.h (hunk 8 above)
Comment (lambdai) [reply source/server/connection_handler_impl.h]: Nice catch!

File: api/envoy/api/v2/listener/listener_components.proto [hunk 11]
```diff
@@ -207,9 +207,32 @@ message FilterChain {
...
 // Listener filter chain match configuration. This is a recursive structure which allows complex
 // nested match configurations to be built using various logical operators.
+//
+// Examples:
+//
+//  disable when the destination port is 3306
+//
+//  code-block:: yaml
+//
+//  destinationPortRange:
```
[the remark this answers is not in the document]
Comment (lambdai) [reply api/envoy/api/v2/listener/listener_components.proto]: Double check later

File: api/envoy/api/v2/listener/listener_components.proto (hunk 5 above)
Comment (lambdai) [reply api/envoy/api/v2/listener/listener_components.proto]: "The filter is enabled when the field is empty"?

File: include/envoy/network/filter.h (hunk 9 above)
Comment (mattklein123) [reply include/envoy/network/filter.h]: Please remove until it's actually needed. Thank you.

Comment (lambdai) [reply include/envoy/network/filter.h]: on it

File: source/common/network/filter_matcher.cc (hunk 10 above)
Comment (lambdai) [reply source/common/network/filter_matcher.cc]: I have different opinions on NOT_REACHED: 
In case if the control plane api has newer version so a new case is provided, NOT_REACHED is hit and the envoy is terminate I suppose.
Throw an exception here allows envoy rejecting the listener config. 
Do you prefer terminate the envoy to received the signal in the aggressive crash way, or assuming envoy will be updated soon while leave the envoy running under previous listener config?

Comment (mattklein123) [reply source/common/network/filter_matcher.cc]: If you have a validator it will throw an exception and the config will be rejected. This is what we do everywhere else.

Comment (lambdai) [reply source/common/network/filter_matcher.cc]: Thanks! I am aware of something like allow-unknown-field but I don't know how could the validator work while we do allow unknown field in this example. 
Could you help me understand if it's a particular validation rule to achieve this, or it's envoy's  default behavior to reject such a config as a-new-matcher in this example?
Meanwhile following your comment to fix

Comment (mattklein123) [reply source/common/network/filter_matcher.cc]: `option (validate.required) = true;` does it which you have.

Comment (lambdai) [reply source/common/network/filter_matcher.cc]: Thanks! I get your point. I didn't know `required` also check if it's *unknown value* ...
Thank you for the information!

File: api/envoy/api/v2/listener/listener_components.proto (hunk 11 above)
[the remark this answers is not in the document]
Comment (lambdai) [reply api/envoy/api/v2/listener/listener_components.proto]: Previewed and it looks better: code-block now acts as a tag

File: source/common/network/filter_matcher.cc (hunk 6 above)
Comment (lambdai) [reply source/common/network/filter_matcher.cc]: Updated: could you take another look?

File: source/common/network/filter_matcher.cc [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+// TwoPhaseMatcher is introduced to save number of memory allocation at build phase memory usage
+// match phase. Work with SetListenerFilterMatcher. The first phase is constructor, and the second
+// phase is finishBuild(). The invariance is that the direct matchers in a set logic matcher should
+// be continuous in the vector.
+//
+// Examples:
+//  `m := A && B && C` where `B := X || Y`
+// The 3 matchers "ABC" should be continuous, while the matchers "XY" should be continuous in the
+// storage of SetListenerFilterMatcher.
+class TwoPhaseMatcher : public ListenerFilterMatcher {
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply source/common/network/filter_matcher.cc]: In thinking about this more, can we just get rid of the two phase concept entirely? I don't think an inline implementation will use much more memory and it will be much easier to understand. The only reason the tap matcher works like this is because it needs to efficiently map a per-request matching state on top of a constant matching config. That doesn't apply here. Thank you!

**Comment by lambdai** [comment]
I will follow your comments. Only one thing:
I think the TwoPhaseMatcher is the leaf of the code. No other users cares about the implementation details here even though there is complexity
And it does save **# of allocations** and could be save more memory if we go through the matcher list to count the number of slots we have.

This implementation is not the hot spot but it's low hanging fruit.

**Comment by mattklein123** [comment]
> This implementation is not the hot spot but it's low hanging fruit.

Please simplify it, thanks. This is premature optimization which we will pay for with making it harder to understand than is needed.

**Comment by lambdai** [comment]
Ack.  Will update PR

File: api/envoy/api/v2/listener/listener_components.proto [hunk 13]
```diff
@@ -207,9 +207,32 @@ message FilterChain {
...
+//
+//  destination_port_range:
+//   start: 3306
+//   end: 3307
+//
+// * Matches if the destination port is 3306 or 15000.
+//
+// .. code-block:: yaml
+//
+//  orMatch:
```
Comment (mattklein123) [inline api/envoy/api/v2/listener/listener_components.proto]: snake_case

File: api/envoy/api/v2/listener/listener_components.proto [hunk 14]
```diff
@@ -255,17 +278,8 @@ message ListenerFilter {
...
-  // 0. always enable filter
-  //    don't set `filter_disabled`
-  // 1. disable when the destination port is 3306
-  //    rule.destination_port_range = Int32Range {start = 3306, end = 3307}
-  // 2. disable when the destination port is 3306 or 15000
-  //    rule.or_match = MatchSet.rules [
-  //       rule.destination_port_range = Int32Range {start = 3306, end = 3307},
-  //       rule.destination_port_range = Int32Range {start = 15000, end = 15001},
-  //    ]
+  // Decide when to disable this listener filter on incoming traffic. Optional match predicate used to disable the filter. The filter is enabled when this field is empty.
```
Comment (mattklein123) [inline api/envoy/api/v2/listener/listener_components.proto]:
```suggestion
  // Optional match predicate used to disable the filter. The filter is enabled when this field is empty.
```

File: source/common/network/filter_matcher.h [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+ * Matcher for implementing set logic.
+ */
+class ListenerFilterSetLogicMatcher : public ListenerFilterMatcher {
+public:
+  explicit ListenerFilterSetLogicMatcher(
+      absl::Span<const ::envoy::config::listener::v3::ListenerFilterChainMatchPredicate* const>
+          predicates);
+
+protected:
+  absl::FixedArray<ListenerFilterMatcherPtr> sub_matchers_;
```
Comment (mattklein123) [inline source/common/network/filter_matcher.h]: I think you probably want InlineArray here. AFAIK FixedArray is typically used for local stack allocations.

**Comment by lambdai** [comment]
./ci/run_envoy_docker.sh './ci/do_ci.sh fix_format' report errors at sample.proto test. 
Looks like the CI on github is passing at that part. Probably my dev env is fishy.

Will address the comment and temporary forget the fix format  pipeline

File: api/envoy/api/v2/listener/listener_components.proto (hunk 13 above)
Comment (lambdai) [reply api/envoy/api/v2/listener/listener_components.proto]: !!! 
sorry for my repeated mistake

File: api/envoy/api/v2/listener/listener_components.proto (hunk 14 above)
Comment (lambdai) [reply api/envoy/api/v2/listener/listener_components.proto]: done

File: source/common/network/filter_matcher.h (hunk 15 above)
Comment (lambdai) [reply source/common/network/filter_matcher.h]: Sorry I don't know which InlineArray you are referring to.
FixedArray beats vector because FixedArray does inline storage when # of element is low. And that is why we use it on stack.
I bet you are not referring to std::array because it require # of element at compile time

Comment (mattklein123) [reply source/common/network/filter_matcher.h]: Sorry I meant https://github.com/abseil/abseil-cpp/blob/master/absl/container/inlined_vector.h.

/wait

**Comment by lambdai** [comment]
Working on the integration test. Thanks!

**Comment by lambdai** [comment]
Integration test added

File: test/config/utility.cc [hunk 16]
```diff
@@ -113,6 +113,13 @@ std::string ConfigHelper::tcpProxyConfig() {
 )EOF");
 }
 
+std::string ConfigHelper::tlsInspectorFilter() {
```
Comment (mattklein123) [inline test/config/utility.cc:116]: cc @lizan who I think added something similar in a current PR and will need a merge.

Comment (lambdai) [reply test/config/utility.cc:116]: This one? https://github.com/envoyproxy/envoy/pull/10569
This PR is on top of the above
